src/FSharp.Compiler.Interactive.Settings/fsihelp.fs

```diff
@@ -0,0 +1,267 @@
+do ()
+
+open System
+open System.Collections.Generic
+open System.IO
+open System.Text
+open System.Reflection
+
+module Parser =
+
+    open System.Xml
+
+    type Help =
+        {
+            Summary: string
+            Remarks: string option
+            Parameters: (string * string) list
+            Returns: string option
+            Exceptions: (string * string) list
+            Examples: (string * string) list
+            FullName: string
+            Assembly: string
+        }
+
+        member this.ToDisplayString() =
+            let sb = StringBuilder()
+
+            let parameters =
+                this.Parameters
+                |> List.map (fun (name, description) -> sprintf "- %s: %s" name description)
+                |> String.concat "\n"
+
+            sb.AppendLine($"\nDescription:\n%s{this.Summary}") |> ignore
+
+            match this.Remarks with
+            | Some r -> sb.AppendLine $"\nRemarks:\n%s{r}" |> ignore
+            | None -> ()
+
+            if not (String.IsNullOrWhiteSpace(parameters)) then
+                sb.AppendLine $"\nParameters:\n%s{parameters}" |> ignore
+
+            match this.Returns with
+            | Some r -> sb.AppendLine $"Returns:\n%s{r}" |> ignore
+            | None -> ()
+
+            if not this.Exceptions.IsEmpty then
+                sb.AppendLine "\nExceptions:" |> ignore
+
+                for (exType, exDesc) in this.Exceptions do
+                    sb.AppendLine $"%s{exType}: %s{exDesc}" |> ignore
+
+            if not this.Examples.IsEmpty then
+                sb.AppendLine "\nExamples:" |> ignore
+
+                for example, desc in this.Examples do
+                    sb.AppendLine example |> ignore
+
+                    if not (String.IsNullOrWhiteSpace(desc)) then
+                        sb.AppendLine $"""// {desc.Replace("\n", "\n// ")}""" |> ignore
+
+                    sb.AppendLine "" |> ignore
+
+            sb.AppendLine $"Full name: %s{this.FullName}" |> ignore
+            sb.AppendLine $"Assembly: %s{this.Assembly}" |> ignore
+
+            sb.ToString()
+
+    let cleanupXmlContent (s: string) = s.Replace("\n ", "\n").Trim() // some stray whitespace from the XML
+
+    // remove any leading `X:` and trailing `N
+    let trimDotNet (s: string) =
+        let s = if s.Length > 2 && s[1] = ':' then s.Substring(2) else s
+        let idx = s.IndexOf('`')
+        let s = if idx > 0 then s.Substring(0, idx) else s
+        s
+
+    let xmlDocCache = Dictionary<string, string>()
+
+    let getXmlDocument xmlPath =
+        match xmlDocCache.TryGetValue(xmlPath) with
+        | true, value ->
+            let xmlDocument = XmlDocument()
+            xmlDocument.LoadXml(value)
+            xmlDocument
+        | _ ->
+            let rawXml = System.IO.File.ReadAllText(xmlPath)
+            let xmlDocument = XmlDocument()
+            xmlDocument.LoadXml(rawXml)
+            xmlDocCache.Add(xmlPath, rawXml)
+            xmlDocument
```

Comment:
Should this just be `tryGetXmlDocument`, with `try Some … with _ -> None`? Then you don't need the `File.Exists` in `getInfos`. In theory the file could disappear or become inaccessible between that call and this one anyway...

Reply:
👍 
Will try tomorrow. It's getting late here...